### PR TITLE
ARROW-9616: [C++] Support LTO for R

### PR DESCRIPTION
**Pull request by nealrichardson**

### CODE REVIEW

**Comment by xhochy** [comment]
@nealrichardson I would rather guess the the issue comes from using the wrong ar/ranlib/nm. There should be a `gcc-ar` / `gcc-nm` / `gcc-ranlib` or so available that can handle the necesary LTO plugin. 

Update: I see @jeroen also runs into the problem in https://github.com/r-windows/rtools-packages/pull/134

File: ci/scripts/PKGBUILD [hunk 1]
```diff
@@ -76,8 +76,10 @@ build() {
     export CPPFLAGS="${CPPFLAGS} -I${MINGW_PREFIX}/include"
     export LIBS="-L${MINGW_PREFIX}/libs"
     export ARROW_S3=OFF
+    export CMAKE_INTERPROCEDURAL_OPTIMIZATION=OFF
   else
     export ARROW_S3=ON
+    export CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON
```
Comment (xhochy) [inline ci/scripts/PKGBUILD:82]:
```suggestion
    export CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON
    export CMAKE_AR=gcc-ar
    export CMAKE_NM=gcc-nm
    export CMAKE_RANLIB=gcc-ranlib
```

**Comment by xhochy** [comment]
Ok, this got us a bit further...

**Comment by xhochy** [comment]
@nealrichardson If you have time to continue this, I guess you need to set `NM=` & co in the environment for `R CMD INSTALL`. I wonder whether the Rtools40 setup should do this automatically, at least for the R sources we don't use CMake.

**Comment by nealrichardson** [comment]
@xhochy that did not seem to work, it still says plugin not found, though maybe I misunderstood what you were suggesting

**Comment by xhochy** [comment]
No, that is what I thought. The logs indicate though that still the wrong tools are used: https://github.com/apache/arrow/runs/1166696816#step:11:180

Maybe this needs to go into `configure.win` or `R/Makevars`. My knowledge of the R build infrastructure is still quite limited 😅

**Comment by nealrichardson** [comment]
🤷  I'll try Makevars

**Comment by nealrichardson** [comment]
No luck. One possibility is that we aren't getting the right 32/64 bit versions of the tools: there are two versions of each, `\rtools40\mingw{32,64}\bin\gcc-{ar,nm,ranlib}`. At least in the most recent attempt, the line with the failure does show mingw64 while it is the 32 bit build and is using the mingw32 compiler.

**Comment by nealrichardson** [comment]
Here's what I'm talking about in the failure in the 32 bit build:

```
lto-wrapper.exe: fatal error: C:\rtools40\mingw32\bin\g++.exe returned 1 exit status
compilation terminated.
C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: error: lto-wrapper failed
collect2.exe: error: ld returned 1 exit status
C:\rtools40\mingw64\bin\nm.exe: D:/a/arrow/arrow/check/arrow.Rcheck/00_pkg_src/arrow/src-i386/array.o: plugin needed to handle lto object
```

Might the problem be that it's trying to use `C:\rtools40\mingw64\bin\nm.exe` instead of `C:\rtools40\mingw32\bin\nm.exe`? I've looked locally in R's etc/Makeconf and don't see where it's configured wrong, but maybe it's buried somewhere.

**Pull request closed without merging**